exp/addons/internal/controllers/clusterresourceset_scope.go

```diff
@@ -0,0 +1,185 @@
+		clusterResourceSet: clusterResourceSet,
+		resourceRef:        resourceRef,
+		resourceSetBinding: resourceSetBinding,
+		data:               normalizedData,
+		normalizedObjs:     objs,
+		computedHash:       computeHash(normalizedData),
+	}
+
+	switch addonsv1.ClusterResourceSetStrategy(clusterResourceSet.Spec.Strategy) {
+	case addonsv1.ClusterResourceSetStrategyApplyOnce:
+		return &reconcileApplyOnceScope{base}
+	case addonsv1.ClusterResourceSetStrategyReconcile:
+		return &reconcileStrategyScope{base}
+	default:
+		return nil
+	}
+}
+
+type baseResourceReconcileScope struct {
+	clusterResourceSet *addonsv1.ClusterResourceSet
+	resourceRef        addonsv1.ResourceRef
+	resourceSetBinding *addonsv1.ResourceSetBinding
+	normalizedObjs     []unstructured.Unstructured
+	data               [][]byte
+	computedHash       string
+}
+
+func (b baseResourceReconcileScope) objs() []unstructured.Unstructured {
+	return b.normalizedObjs
+}
+
+func (b baseResourceReconcileScope) hash() string {
+	return b.computedHash
+}
+
+type reconcileStrategyScope struct {
+	baseResourceReconcileScope
+}
+
+func (r *reconcileStrategyScope) needsApply() bool {
+	resourceBinding := r.resourceSetBinding.GetResource(r.resourceRef)
+
+	return resourceBinding == nil || !resourceBinding.Applied || resourceBinding.Hash != r.computedHash
+}
+
+func (r *reconcileStrategyScope) apply(ctx context.Context, c client.Client) error {
+	return apply(ctx, c, r.applyObj, r.objs())
+}
+
+func (r *reconcileStrategyScope) applyObj(ctx context.Context, c client.Client, obj *unstructured.Unstructured) error {
+	currentObj := &unstructured.Unstructured{}
+	currentObj.SetAPIVersion(obj.GetAPIVersion())
+	currentObj.SetKind(obj.GetKind())
+	err := c.Get(ctx, client.ObjectKeyFromObject(obj), currentObj)
+	if apierrors.IsNotFound(err) {
+		return createUnstructured(ctx, c, obj)
+	}
+	if err != nil {
+		return errors.Wrapf(
+			err,
+			"reading object %s %s",
+			obj.GroupVersionKind(),
+			klog.KObj(obj),
+		)
+	}
+
+	patch := client.MergeFrom(currentObj.DeepCopy())
+	obj.SetResourceVersion(currentObj.GetResourceVersion())
+	if err = c.Patch(ctx, obj, patch); err != nil {
+		return errors.Wrapf(
+			err,
+			"patching object %s %s/%s",
+			obj.GroupVersionKind(),
+			obj.GetNamespace(),
+			obj.GetName(),
+		)
+	}
+
+	return nil
+}
+
+type reconcileApplyOnceScope struct {
+	baseResourceReconcileScope
+}
+
+func (r *reconcileApplyOnceScope) needsApply() bool {
+	return !r.resourceSetBinding.IsApplied(r.resourceRef)
+}
+
+func (r *reconcileApplyOnceScope) apply(ctx context.Context, c client.Client) error {
+	return apply(ctx, c, r.applyObj, r.objs())
+}
+
+func (r *reconcileApplyOnceScope) applyObj(ctx context.Context, c client.Client, obj *unstructured.Unstructured) error {
+	// The create call is idempotent, so if the object already exists
+	// then do not consider it to be an error.
+	if err := createUnstructured(ctx, c, obj); !apierrors.IsAlreadyExists(err) {
+		return err
+	}
```

Comment:
```suggestion
        if err := createUnstructured(ctx, c, obj); err != nil && !apierrors.IsAlreadyExists(err) {
		return err
	}
```
nit: If we keep it on this level, can we please also check for nil.

I know the result is the same as `apierrors.IsAlreadyExists` returns `false` for nil and then `return err` returns nil because err is nil.

Just seems more obvious if we just check for nil directly

Reply:
Fair enough, done